### PR TITLE
listener: envoy.reloadable_features.listener_reuse_port_default_enabled deprecation

### PR DESCRIPTION
**Pull request by giantcroc**
Signed-off-by: changran <changran.wang@intel.com>

Commit Message: remove envoy.reloadable_features.listener_reuse_port_default_enabled
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/server/server.cc
Comment (KfreeZ) [inline source/server/server.cc:533]: from the 1.20.oc doc, the default value is set to false, please correct me if I'm wrong.
ref to v1.20.0.rst "```
To retain the previous behavior, either explicitly set the new configuration
  field to false, or set the runtime feature flag ``envoy.reloadable_features.listener_reuse_port_default_enabled``
  to false.
```"

Comment (giantcroc) [reply source/server/server.cc:533]: I think the default value is true from my experience, and this feature is enabled by default.

**Comment by KBaichoo** [comment]
/assign @mattklein123 

It seems unclear if we want to remove this now as this might not yet be resolved?
> As part of this change, the use of reuse_port for TCP listeners on both macOS and Windows has been disabled due to suboptimal behavior. See the field documentation for more information.

from https://www.envoyproxy.io/docs/envoy/latest/version_history/v1.20.0

File: source/server/server.cc
Comment (mattklein123) [reply source/server/server.cc:533]: Correct, the new default is true.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
I think it's OK to remove the runtime flag and just leave the default as disabled for OSX and Windows. Please merge main to fix CI.

/wait

File: source/server/server.cc
Comment (soulxu) [inline source/server/server.cc:1014]: Is that means we only have `true` and `false` value for the `enable_reuse_port_default_`? so `enable_reuse_port_default_` can be a normal bool instead of the enum `ReusePortDefault`, and the enum `ReusePortDefault` becomes useless.

File: source/server/server.cc
Comment (soulxu) [inline source/server/server.cc:530]: A question, does it mean the user never can upgrade to the true default value by hot restart? So that means we should keep those code forever?

File: test/server/server_test.cc
Comment (soulxu) [inline test/server/server_test.cc:1029]: Emm...this test doesn't seem a specific test for `server_->enableReusePortDefault()`, so I don't know if we should delete it directly.

**Comment by daixiang0** [comment]
Please run `ci/run_envoy_docker.sh 'ci/check_and_fix_format.sh'` to fix CI.

File: source/server/server.cc
Comment (soulxu) [inline source/server/server.cc:532]: It is already disabled by here

https://github.com/envoyproxy/envoy/blob/158535979a833d076c44c701f1cdf5406187d3ba/source/server/listener_impl.cc#L920-L931

So I guess we needn't disable it here again

File: source/server/server.cc
Comment (giantcroc) [reply source/server/server.cc:1014]: > Is that means we only have `true` and `false` value for the `enable_reuse_port_default_`? so `enable_reuse_port_default_` can be a normal bool instead of the enum `ReusePortDefault`, and the enum `ReusePortDefault` becomes useless.

Yeah, I think it's good to remove the enum `ReusePortDefault`.

File: source/server/server.cc
Comment (giantcroc) [reply source/server/server.cc]: OK,thanks.

File: source/server/server.h
Comment (soulxu) [inline source/server/server.h:307]: This can be removed also.

File: test/server/server_test.cc
Comment (soulxu) [inline test/server/server_test.cc:1040]: ditto, also don't think this test is `server_->enableReusePortDefault()` specific

**Review by soulxu — CHANGES_REQUESTED** [review]

**Comment by mattklein123** [comment]
LGTM pending @soulxu comments, thank you.

/wait

File: source/server/server.cc
Comment (soulxu) [reply source/server/server.cc]: @mattklein123 this question probably needs your help. I thought we should remove all the upgrade code. But from the code and comments, I understand the hot restart can't change the default value. That means we only can remove those codes when all the users have a cold restart. Then probably this code never can be deleted?

Comment (mattklein123) [reply source/server/server.cc]: We don't have a lot of precedent for this. I think technically we could likely remove the code when the last version that had the old default falls out of support. You could leave a comment/TODO about this and we can circle back later? cc @envoyproxy/envoy-maintainers

File: source/server/server.h
Comment (giantcroc) [reply source/server/server.h:307]: > Nobody is using this anymore, so we can remove it.

OK, thanks.

File: test/server/server_test.cc
Comment (giantcroc) [reply test/server/server_test.cc:1040]: > ditto, also don't think this test is `server_->enableReusePortDefault()` specific.

Yeah, I also think these tests should  be kept, and just removing the runtime feature setting in runtime_bootstrap.

**Comment by soulxu** [comment]
LGTM, thanks for the update!

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**
